SimCalorimetry/HGCalSimProducers/src/HGCDigitizer.cc

```diff
@@ -236,35 +371,46 @@ void HGCDigitizer::initializeEvent(edm::Event const& e, edm::EventSetup const& e
   // reserve memory for a full detector
   unsigned idx = getType();
   simHitAccumulator_->reserve(averageOccupancies_[idx] * validIds_.size());
+  pusimHitAccumulator_->reserve(averageOccupancies_[idx] * validIds_.size());
+  //hitOrder_monitor->reserve( validIds_.size());
 }
 
 //
 void HGCDigitizer::finalizeEvent(edm::Event& e, edm::EventSetup const& es, CLHEP::HepRandomEngine* hre) {
   hitRefs_bx0.clear();
-
+  hitOrder_monitor.clear();
   const CaloSubdetectorGeometry* theGeom = (nullptr == gHGCal_ ? static_cast<const CaloSubdetectorGeometry*>(gHcal_)
                                                                : static_cast<const CaloSubdetectorGeometry*>(gHGCal_));
 
   ++nEvents_;
+
   unsigned idx = getType();
   // release memory for unfilled parts of hash table
   if (validIds_.size() * averageOccupancies_[idx] > simHitAccumulator_->size()) {
     simHitAccumulator_->reserve(simHitAccumulator_->size());
+    pusimHitAccumulator_->reserve(simHitAccumulator_->size());
   }
   //update occupancy guess
   const double thisOcc = simHitAccumulator_->size() / ((double)validIds_.size());
   averageOccupancies_[idx] = (averageOccupancies_[idx] * (nEvents_ - 1) + thisOcc) / nEvents_;
 
   if (premixStage1_) {
-    auto simResult = std::make_unique<PHGCSimAccumulator>();
-    if (!simHitAccumulator_->empty()) {
-      saveSimHitAccumulator(*simResult, *simHitAccumulator_, validIds_, premixStage1MinCharge_, premixStage1MaxCharge_);
+    std::unique_ptr<PreMixHGCSimAccumulator> simRecord;
+
+    if (!pusimHitAccumulator_->empty()) {
+      simRecord = std::make_unique<PreMixHGCSimAccumulator>();
```

Comment:
Please keep the object always created
```suggestion
    auto simRecord = std::make_unique<PreMixHGCSimAccumulator>()

    if (!pusimHitAccumulator_->empty()) {
```

Reply:
If I follow your suggestion, then what to do here under the "if " condition in line 401?

Reply:
Leave the call to `saveSimHitAccumulator_forPreMix()`  there. On the other hand, the function itself looks like it would work with empty input, so one option could be to remove the "if" altogether.

Reply:
I would like to hold off on this suggestion for now. But I will keep it in mind and see if I can do away with if condition in the next PR.

Reply:
Just to make sure, the `PreMixHGCSimAccumulator` must be created for each event (not doing it was an oversight from me and was fixed recently after causing problems elsewhere, we should avoid to fix it again). The question of "if" can be left to later PR.

Reply:
This one still needs to be fixed.